### PR TITLE
Load based on existing table schema

### PR DESCRIPTION
**Pull request by jp-harvey**

### CODE REVIEW

File: pymapd/_pandas_loaders.py [hunk 1]
```diff
@@ -79,35 +80,36 @@ def thrift_cast(data, mapd_type):
...
         df = df.reset_index()
 
     input_cols = []
-    all_nulls = None
-
-    for col in df.columns:
+    for col, mapd_type in tbl_cols.items():
         data = df[col]
-        mapd_type = get_mapd_dtype(data)
+        print(mapd_type)
```
Comment (andrewseidl) [inline pymapd/_pandas_loaders.py]: :speak_no_evil:

File: tests/test_loaders.py [hunk 2]
```diff
@@ -38,6 +41,19 @@ def test_build_table_columnar_pandas(self):
         import pandas as pd
         import numpy as np
 
+        columns = {
+            'boolean_': 'BOOL',
+            'smallint_': 'TINYINT',
```
Comment (andrewseidl) [inline tests/test_loaders.py]: `smallint_` should be a `SMALLINT`. Should add a `TINYINT` to the test.

Comment (jp-harvey) [reply tests/test_loaders.py]: Done

**Comment by randyzwitch** [comment]
@jp-harvey This PR is starting to get really complicated...can you split things out that seem like bugs and improving the tests from the column ordering change?

**Comment by randyzwitch** [comment]
@jp-harvey Can you split this PR again? I feel like we really haven’t agreed that fixing the ordered dict/pandas bug is something we should do within pymapd, while chunked data loading is probably a good idea in its own right

**Comment by jp-harvey** [comment]
@randyzwitch yes if absolutely necessary I can. The column ordering is not necessarily a fix, although it was originally logged that way, it's actually a feature. It comes with the added benefit of not having to detect the column types which can be pretty gnarly in some cases, one example is when there are no values in a column it's not possible to import the dataframe because it has no values to detect the type.

The previous way still works as it used to and this is optional so the question we should probably be asking is why would we _not_ do it?

**Comment by randyzwitch** [comment]
> The previous way still works as it used to and this is optional so the question we should probably be asking is why would we not do it?

Overall, my concern is that it seems brittle to try and align columns based on their name, when the python/pandas name doesn't follow the same rules as OmniSci.

While your example is from hand-coding dicts, I'm concerned that this functionality will silently misalign the columns in unexpected ways, similar to a bug I found in the JDBC driver which silently puts the columns in alphabetical order. 

For example, a data frame could be defined with columns `test`, `Test`, and `Test `, which would be valid in pandas but not valid in OmniSci. Because ultimately our data loader is positional, we're either going to throw an error in pymapd or decide one of the columns is the correct one and load it. Which seems like too much magic to me. 

So this functionality in the best case scenario works by "do what I mean". My preference would be to have less magic in pymapd (this includes the cascading `load_table` method currently deciding to load rowwise, columnwise or arrow) and instead of requiring users to make sure their data actually conforms to the existing table definition.

So the trade-off I see is using ordered dicts for your use-case to preserve insertion order vs. adding another branch of logic to debug for users into the future.

The other part about splitting the PRs is that you're doing 3 things here:

- Changing the type inference (a breaking change)
- Column ordering (new feature)
- Chunking (new feature)

While the type inference stuff makes logical sense, I'm concerned that this could change real-world code in ways that our limited test suite may not pick up. So it'd be great to split each of these features out, so that we can cleanly revert them in the future if needed.

**Comment by jp-harvey** [comment]
> While your example is from hand-coding dicts, I'm concerned that this functionality will silently misalign the columns in unexpected ways, similar to a bug I found in the JDBC driver which silently puts the columns in alphabetical order.

Either way the user needs to make a choice - either order them right or name them right. I'm nowhere near the level of experience as you in these types of things so maybe there's something I just don't get but I don't see why one option is better than the other so why not have the choice? My main use case is extracting data via APIs which always have a column name but may or may not be ordered. Typically then I'll extract what I need for insert. I already have to know the name in the dict and my table, why do I now have to do an extra step and order the columns correctly as well? And what's the technical purity around column ordering as opposed to names, when in human terms the order of the columns is almost exclusively going to be secondary to the names?

> For example, a data frame could be defined with columns test, Test, and Test , which would be valid in pandas but not valid in OmniSci. Because ultimately our data loader is positional, we're either going to throw an error in pymapd or decide one of the columns is the correct one and load it. Which seems like too much magic to me.

Yes, using PyMapD is already like trying to convince a petulant child to tell you where they hid your phone that they've stolen, it blows a raspberry at you with some obscure error message (In fairness it's often Thrift or OmniSci server where the actual problem lies, but the point remains). From a user perspective why is one obscure error better than another obscure error? Why can't we add functionality to allow people to realize "oh, RTFM I just need to pass that flag and it gets sorted out" instead of "here goes 2 hours of my day in a rabbit hole"? Our mission contains "...for everyone" not "...for everyone as long as they understand all the nuances of our software and compensate for those in their own code". Why do people need to care that our data load is positional? 

With your specific example, we could easily throw a meaningful error about duplicate column names - a far more meaningful error than the obscure Thrift / OmniSci ones that tend to come back. I'm not putting that in right now though, that can be a later improvement. We could think forever on every possibility but that's not progress.

The column types is actually as much or more of an argument for this than the column names (and perhaps could be split out as its own feature). The fact that Python / Pandas and OmniSci types don't match up speaks to your point about the guessing PyMapD does already. Using the table schema is unambiguous, unlike the current method you have concerns with, and opens up the ability to convert to the actual type (which has implications for nulls at the very least) instead of using a best fit default in ambiguous situations.

I will split these out into different PRs, that is logical. The resistance to this particular schema matching feature is starting to feel like we're terrified of making fixable mistakes in software that's already far from perfect, which means we're not making improvements. Let's get some progress on making our tools easier to use for less sophisticated people (ie. like me), even if it's not perfect.

**Comment by jp-harvey** [comment]
Closing this, there have been other fixes related to this PR, plus there should be the option of matching the types without having to match the names.

**Pull request closed without merging**